### ISSUE TITLE
Get metrics from SQL Database (PaaS, ARM)

### ISSUE THREAD

**Issue by Ninroot** [Azure/azure-sdk-for-python#909]
I am trying to get metrics from SQL Database Azure (PaaS, ARM). I followed the example given on the [documentation](http://azure-sdk-for-python.readthedocs.io/en/latest/sample_azure-monitor.html) and adapted from SQL Database. The program returns the good definitions I found on the [Azure documentation](https://docs.microsoft.com/en-in/azure/monitoring-and-diagnostics/monitoring-supported-metrics#microsoftsqlserversdatabases) but the values are None. Is the SDK supposed to work in this case ?

Code to get the storage percent :
```
def get_metrics_API(client, subscription_id, resource_group_name, server_name, db_name):
    resource_id = (
        "subscriptions/{}/"
        "resourceGroups/{}/"
        "providers/Microsoft.Sql/servers/{}/"
        "databases/{}"  
    ).format(subscription_id, resource_group_name, server_name, db_name)

     You can get the available metrics of this specific resource
    for metric in client.metric_definitions.list(resource_id):
         azure.monitor.models.MetricDefinition
        print("{}: id={}, unit={}".format(
            metric.name.localized_value,
            metric.name.value,
            metric.unit
        ))

    today = datetime.datetime.now().date()
    yesterday = today - datetime.timedelta(days=1)

    filter = " and ".join([
        "name.value eq 'storage_percent'",
        "aggregationType eq 'Maximum'",
        "startTime eq {}".format(yesterday),
        "endTime eq {}".format(today),
        #"timeGrain eq 'PT1H'" # duration'PT1H'
    ])

    metrics_data = client.metrics.list(
        resource_id,
        # filter=filter,
        # raw=True
    )

    for item in metrics_data:
        # azure.monitor.models.Metric
        print("{} ({})".format(item.name.localized_value, item.unit.name))
        for data in item.data:
            # azure.monitor.models.MetricData
            print("{}: {}".format(data.time_stamp, data))
```

Console :
```
CPU percentage: id=cpu_percent, unit=Unit.percent
Data IO percentage: id=physical_data_read_percent, unit=Unit.percent
Log IO percentage: id=log_write_percent, unit=Unit.percent
DTU percentage: id=dtu_consumption_percent, unit=Unit.percent
Total database size: id=storage, unit=Unit.bytes
Successful Connections: id=connection_successful, unit=Unit.count
Failed Connections: id=connection_failed, unit=Unit.count
Blocked by Firewall: id=blocked_by_firewall, unit=Unit.count
Deadlocks: id=deadlock, unit=Unit.count
Database size percentage: id=storage_percent, unit=Unit.percent
In-Memory OLTP storage percent(Preview): id=xtp_storage_percent, unit=Unit.percent
Workers percentage: id=workers_percent, unit=Unit.percent
Sessions percent: id=sessions_percent, unit=Unit.percent
DTU limit: id=dtu_limit, unit=Unit.count
DTU used: id=dtu_used, unit=Unit.count
CPU percentage (percent)
2016-12-07 12:42:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2016, 12, 7, 12, 42, tzinfo=<isodate.tzinfo.Utc object at 0x7f242fb3b5d0>), 'total': None}
2016-12-07 12:43:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2016, 12, 7, 12, 43, tzinfo=<isodate.tzinfo.Utc object at 0x7f242fb3b5d0>), 'total': None}
...
```

**Comment by lmazuel** [comment]
Hi @Ninroot 

First, your query is asking for aggregation Maximum, so you will have a value only in the `maximum` attribute. Second, I'm pretty sure the timeGrain has to be inside the query (you removed it). The [MetricDefinition](http://azure-sdk-for-python.readthedocs.io/en/latest/ref/azure.monitor.models.html#azure.monitor.models.MetricDefinition) object you have will give you a `metric_availabilities` attribute with all possible timegrain.

If you could check these points, would be helpful. In the meantime, I will try to test myself on a SQL DB. If I found the same problem, I will check the Python code and at worst will contact the Monitor RestAPI team for guidance.

Thank you,

**Comment by Ninroot** [comment]
Hi @lmazuel,

> First, your query is asking for aggregation Maximum, so you will have a value only in the maximum attribute. 
When I print `data.maximum `I still have `None` value.

> Second, I'm pretty sure the timeGrain has to be inside the query (you removed it).
I still have `None` value with `"timeGrain eq 'PT1H'"` or without it in my filter. (is that a correct value ?)

Thank you

**Comment by lmazuel** [comment]
Hi @Ninroot ,

Sorry for the late answer. To know if PT1H is correct value in your context, please call `metric_availabilities` using your resource id.

I'll take a look asap next week, hopefully with vacation time for a lot of people I will find time more easily :)

**Comment by lmazuel** [comment]
Hi @Ninroot 
I was able to reproduce the issue, it's not from Python but from the RestAPI. I asked directly the Monitor team. I'll let you know when I receive an answer.
Thank you,

**Comment by Ninroot** [comment]
Hi @lmazuel ,
Thank you for your feedback.

**Comment by lmazuel** [comment]
Hi @Ninroot 

I got feedback from the SQL team directly. When a DB is not used for a long time, metrics RestAPI answers an empty payload (translated into `None` in Python). They are aware that this can be workarounded for CPU usage (empty means 0) but not for DB size and some metrics like this one. So the RestAPI will be fixed soon to have data anytime, even if the DB is unused. It will be fixed in Python automatically, no need for a new Python release.

I tested myself last week on an old DB unused and it's why I reproduced the issue. I made some queries on it this morning, to remove the "unused" state and it's ok. See an example of pyaload I just got (I used DB at 18:04 UTC today):
```json
{"timeStamp":"2017-01-04T17:00:00Z"},{"timeStamp":"2017-01-04T18:00:00Z","total":0.0}
```

Could you confirm you were in a situation of unused DB for a few days?

**Comment by Ninroot** [comment]
Hi @lmazuel,
Okay, thank you! I still think `None` is a weird value instead of 0, I would use it for specifics cases like asking for 12h old values with a database created 1h ago.
Anyway, I have `None` for the `storage_percent` metrics for empty database (created 3h ago).
I also have `None` for the `storage_percent` metrics for **not** empty database (created 3h ago too) : with AdventureWorksLS example.
In both case I didn't interact with the database (insert/delete or even select) but the size given by Azure portal is not null (around 12Mo).

**Comment by lmazuel** [comment]
Could you just try to connect to it (no interaction; just connection) and tell me what you got with metrics?

**Comment by Ninroot** [comment]
Hi @lmazuel,
I think you had a good feeling because I got something after a connexion.

storage_percent on fresh empty sql server database **without connection** :
```
2017-01-06 10:00:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 0, tzinfo=<isodate.tzinfo.Utc object at 0x7fc04c687d50>), 'total': None}
2017-01-06 10:01:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 1, tzinfo=<isodate.tzinfo.Utc object at 0x7fc04c687d50>), 'total': None}
2017-01-06 10:02:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 2, tzinfo=<isodate.tzinfo.Utc object at 0x7fc04c687d50>), 'total': None}
2017-01-06 10:03:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 3, tzinfo=<isodate.tzinfo.Utc object at 0x7fc04c687d50>), 'total': None}
2017-01-06 10:04:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 4, tzinfo=<isodate.tzinfo.Utc object at 0x7fc04c687d50>), 'total': None}
2017-01-06 10:05:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 5, tzinfo=<isodate.tzinfo.Utc object at 0x7fc04c687d50>), 'total': None}
2017-01-06 10:06:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 6, tzinfo=<isodate.tzinfo.Utc object at 0x7fc04c687d50>), 'total': None}
2017-01-06 10:07:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 7, tzinfo=<isodate.tzinfo.Utc object at 0x7fc04c687d50>), 'total': None}
2017-01-06 10:08:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 8, tzinfo=<isodate.tzinfo.Utc object at 0x7fc04c687d50>), 'total': None}
2017-01-06 10:09:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 9, tzinfo=<isodate.tzinfo.Utc object at 0x7fc04c687d50>), 'total': None}
2017-01-06 10:10:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 10, tzinfo=<isodate.tzinfo.Utc object at 0x7fc04c687d50>), 'total': None}
2017-01-06 10:11:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 11, tzinfo=<isodate.tzinfo.Utc object at 0x7fc04c687d50>), 'total': None}
2017-01-06 10:12:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 12, tzinfo=<isodate.tzinfo.Utc object at 0x7fc04c687d50>), 'total': None}
2017-01-06 10:13:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 13, tzinfo=<isodate.tzinfo.Utc object at 0x7fc04c687d50>), 'total': None}
2017-01-06 10:14:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 14, tzinfo=<isodate.tzinfo.Utc object at 0x7fc04c687d50>), 'total': None}
2017-01-06 10:15:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 15, tzinfo=<isodate.tzinfo.Utc object at 0x7fc04c687d50>), 'total': None}
```

storage_percent on fresh empty sql server database **after simple connection** with visual studio :
```
2017-01-06 10:00:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 0, tzinfo=<isodate.tzinfo.Utc object at 0x7f57f60a7d50>), 'total': None}
2017-01-06 10:01:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 1, tzinfo=<isodate.tzinfo.Utc object at 0x7f57f60a7d50>), 'total': None}
2017-01-06 10:02:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 2, tzinfo=<isodate.tzinfo.Utc object at 0x7f57f60a7d50>), 'total': None}
2017-01-06 10:03:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 3, tzinfo=<isodate.tzinfo.Utc object at 0x7f57f60a7d50>), 'total': None}
2017-01-06 10:04:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 4, tzinfo=<isodate.tzinfo.Utc object at 0x7f57f60a7d50>), 'total': None}
2017-01-06 10:05:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 5, tzinfo=<isodate.tzinfo.Utc object at 0x7f57f60a7d50>), 'total': None}
2017-01-06 10:06:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 6, tzinfo=<isodate.tzinfo.Utc object at 0x7f57f60a7d50>), 'total': None}
2017-01-06 10:07:00+00:00: {'count': None, 'average': None, 'maximum': 0.0, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 7, tzinfo=<isodate.tzinfo.Utc object at 0x7f57f60a7d50>), 'total': None}
2017-01-06 10:08:00+00:00: {'count': None, 'average': None, 'maximum': 0.0, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 8, tzinfo=<isodate.tzinfo.Utc object at 0x7f57f60a7d50>), 'total': None}
2017-01-06 10:09:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 9, tzinfo=<isodate.tzinfo.Utc object at 0x7f57f60a7d50>), 'total': None}
2017-01-06 10:10:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 10, tzinfo=<isodate.tzinfo.Utc object at 0x7f57f60a7d50>), 'total': None}
2017-01-06 10:11:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 11, tzinfo=<isodate.tzinfo.Utc object at 0x7f57f60a7d50>), 'total': None}
2017-01-06 10:12:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 12, tzinfo=<isodate.tzinfo.Utc object at 0x7f57f60a7d50>), 'total': None}
2017-01-06 10:13:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 13, tzinfo=<isodate.tzinfo.Utc object at 0x7f57f60a7d50>), 'total': None}
2017-01-06 10:14:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 14, tzinfo=<isodate.tzinfo.Utc object at 0x7f57f60a7d50>), 'total': None}
2017-01-06 10:15:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 15, tzinfo=<isodate.tzinfo.Utc object at 0x7f57f60a7d50>), 'total': None}
```

cpu_percent on fresh empty sql server database after simple connection with visual studio :
```
2017-01-06 10:00:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 0, tzinfo=<isodate.tzinfo.Utc object at 0x7f94d7dc4d90>), 'total': None}
2017-01-06 10:01:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 1, tzinfo=<isodate.tzinfo.Utc object at 0x7f94d7dc4d90>), 'total': None}
2017-01-06 10:02:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 2, tzinfo=<isodate.tzinfo.Utc object at 0x7f94d7dc4d90>), 'total': None}
2017-01-06 10:03:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 3, tzinfo=<isodate.tzinfo.Utc object at 0x7f94d7dc4d90>), 'total': None}
2017-01-06 10:04:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 4, tzinfo=<isodate.tzinfo.Utc object at 0x7f94d7dc4d90>), 'total': None}
2017-01-06 10:05:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 5, tzinfo=<isodate.tzinfo.Utc object at 0x7f94d7dc4d90>), 'total': None}
2017-01-06 10:06:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 6, tzinfo=<isodate.tzinfo.Utc object at 0x7f94d7dc4d90>), 'total': None}
2017-01-06 10:07:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 7, tzinfo=<isodate.tzinfo.Utc object at 0x7f94d7dc4d90>), 'total': None}
2017-01-06 10:08:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 8, tzinfo=<isodate.tzinfo.Utc object at 0x7f94d7dc4d90>), 'total': None}
2017-01-06 10:09:00+00:00: {'count': None, 'average': None, 'maximum': 10.634, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 9, tzinfo=<isodate.tzinfo.Utc object at 0x7f94d7dc4d90>), 'total': None}
2017-01-06 10:10:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 10, tzinfo=<isodate.tzinfo.Utc object at 0x7f94d7dc4d90>), 'total': None}
2017-01-06 10:11:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 11, tzinfo=<isodate.tzinfo.Utc object at 0x7f94d7dc4d90>), 'total': None}
2017-01-06 10:12:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 12, tzinfo=<isodate.tzinfo.Utc object at 0x7f94d7dc4d90>), 'total': None}
2017-01-06 10:13:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 13, tzinfo=<isodate.tzinfo.Utc object at 0x7f94d7dc4d90>), 'total': None}
2017-01-06 10:14:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 14, tzinfo=<isodate.tzinfo.Utc object at 0x7f94d7dc4d90>), 'total': None}
2017-01-06 10:15:00+00:00: {'count': None, 'average': None, 'maximum': None, 'minimum': None, 'time_stamp': datetime.datetime(2017, 1, 6, 10, 15, tzinfo=<isodate.tzinfo.Utc object at 0x7f94d7dc4d90>), 'total': None}
```

**Comment by lmazuel** [comment]
@Ninroot I just checked with the SQL team again, they will deploy a fix soon (next month likely) that will fill values all the time. In the meantime, you have to consider that `None` is equivalent to "0.0".

I keep this issue opened until I see the RestAPI in production.

**Comment by Ninroot** [comment]
@lmazuel, okay thank you 👍

**Comment by lmazuel** [comment]
Hi @Ninroot 
The fix is supposed to be online. Indeed, I don't see the issue anymore, could you confirm on your side?

**Comment by lmazuel** [comment]
I close this issue due to inactivity and because I think it's solved anyway. Feel free to ask me to re-open it or open a new issue if necessary.
Thank you!
